text/0000-const-generics.md

```diff
@@ -0,0 +1,287 @@
+parameters is restricted later in this RFC.
+
+The const parameter is in scope for the entire body of the item (type, impl,
+function, method, etc) in which it is declared.
+
+## Applying a const as a parameter
+
+Any const expression of the type ascribed to a const parameter can be applied
+as that parameter. When applying an expression as const parameter (except for
+arrays), which is not an identity expression, the expression must be contained
+within a block. This syntactic restriction is necessary to avoid requiring
+infinite lookahead when parsing an expression inside of a type.
+
+```rust
+const X: usize = 7;
+
+let x: RectangularArray<i32, 2, 4>;
+let y: RectangularArray<i32, X, {2 * 2}>;
+```
+
+### Arrays
+Arrays have a special construction syntax: `[T; CONST]`. In array syntax,
+braces are not needed around any const expressions; `[i32; N * 2]` is a
+syntactically valid type.
+
+## When a const variable can be used
+
+A const variable can be used as a const in any of these contexts:
+
+1. As an applied const to any type which forms a part of the signature of
+the item in question: `fn foo<const N: usize>(arr: [i32; N])`.
+2. As part of a const expression used to define an associated const, or as a
+parameter to an associated type.
+3. As a value in any runtime expression in the body of any functions in the
+item.
+4. As a parameter to any type used in the body of any functions in the item,
+as in `let x: [i32; N]` or `<[i32; N] as Foo>::bar()`.
+5. As a part of the type of any fields in the item (as in
+`struct Foo<const N: usize>([i32; N]);`).
+
+In general, a const variable can be used where a const can. There is one
+significant exception: const variables cannot be used in the construction of
+consts, statics, functions, or types inside a function body. That is, these
+are invalid:
+
+```rust
+fn foo<const X: usize>() {
+    const Y: usize = X * 2;
+    static Z: (usize, usize)= (X, X);
+
+    struct Foo([i32; X]);
+}
+```
+
+This restriction can be analogized to the restriction on using type variables
+in types constructed in the body of functions - all of these declarations,
+though private to this item, must be independent of it, and do not have any
+of its parameters in scope.
+
+## Theory of equality for type equality of two consts
+
+During unification and the overlap check, it is essential to determine when two
+types are equivalent or not. Because types can now be dependent on consts, we
+must define how we will compare the equality of two constant expressions.
+
+For most cases, the equality of two consts follows the same reasoning you would
+expect - two constant values are equal if they are equal to one another. But
+there are some particular caveats.
+
+### Structural equality
+
+Const equality is determined according to the definition of structural equality
+defined in [RFC 1445][1445]. Only types which have the "structural match"
+property can be used as const parameters. This would exclude floats, for
+example.
```

Comment:
C++ supports reference const parameters and uses "structural" bitwise comparison to unify them.
http://coliru.stacked-crooked.com/a/3301c82ba77a2f32